src/controllers/controllerhidreporttabsmanager.cpp

```diff
@@ -0,0 +1,488 @@
+        if (item) {
+            // Retrieve custom data from the first cell
+            QVariant customData = table->item(row, 0)->data(Qt::UserRole + 1);
+            if (customData.isValid()) {
+                auto control =
+                        reinterpret_cast<hid::reportDescriptor::Control*>(
+                                customData.value<void*>());
+                // Set the control value in the reportData
+                bool success = hid::reportDescriptor::applyLogicalValue(
+                        reportData, *control, item->text().toLongLong());
+                if (!success) {
+                    qWarning() << "Failed to set control value for row" << row;
+                    continue;
+                }
+            }
+        }
+    }
+
+    // Send the reportData
+    if (reportType == hid::reportDescriptor::HidReportType::Feature) {
+        jsProxy->sendFeatureReport(reportId, reportData);
+    } else if (reportType == hid::reportDescriptor::HidReportType::Output) {
+        jsProxy->sendOutputReport(reportId, reportData);
+    }
+}
+
+void ControllerHidReportTabsManager::populateHidReportTable(
+        QTableWidget* table,
+        const hid::reportDescriptor::Report& report,
+        hid::reportDescriptor::HidReportType reportType) {
+    // Temporarily disable updates to speed up populating
+    table->setUpdatesEnabled(false);
+
+    // Reserve rows up-front
+    const auto& controls = report.getControls();
+    table->setRowCount(static_cast<int>(controls.size()));
+
+    // Set the delegate once if needed
+    if (reportType != hid::reportDescriptor::HidReportType::Input) {
+        table->setItemDelegateForColumn(5, new ValueItemDelegate(table));
+    }
+
+    bool showVolatileColumn = (reportType == hid::reportDescriptor::HidReportType::Feature ||
+            reportType == hid::reportDescriptor::HidReportType::Output);
+
+    // Set headers
+    QStringList headers = {QStringLiteral("Byte Position"),
+            QStringLiteral("Bit Position"),
+            QStringLiteral("Bit Size"),
+            QStringLiteral("Logical Min"),
+            QStringLiteral("Logical Max"),
+            QStringLiteral("Value"),
+            QStringLiteral("Physical Min"),
+            QStringLiteral("Physical Max"),
+            QStringLiteral("Unit Scaling"),
+            QStringLiteral("Unit"),
+            QStringLiteral("Abs/Rel"),
+            QStringLiteral("Wrap"),
+            QStringLiteral("Linear"),
+            QStringLiteral("Preferred"),
+            QStringLiteral("Null")};
+    if (showVolatileColumn) {
+        headers << QStringLiteral("Volatile");
+    }
+    headers << QStringLiteral("Usage Page") << QStringLiteral("Usage");
+
+    table->setColumnCount(headers.size());
+    table->setHorizontalHeaderLabels(headers);
+    table->verticalHeader()->setVisible(false);
+
+    // Helpers
+    auto createReadOnlyItem = [](const QString& text, bool rightAlign = false) {
```

Comment:
Is it possible to move that into an anonymous namespace? The same below. This is the prefered way to define free named functions.